bql/semantic/expression_test.go

```diff
@@ -496,20 +592,150 @@ func TestNewEvaluator(t *testing.T) {
+			},
+			err:  false,
+			want: true,
+		},
+		{
+			id: `?o > "37"^^type:int64`,
+			in: []ConsumedElement{
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemBinding,
+					Text: "?o",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemGT,
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemLiteral,
+					Text: `"37"^^type:int64`,
+				}),
+			},
+			r: table.Row{
+				"?o": &table.Cell{N: newNodeFromStringOrDie("/u", "paul")},
+			},
+			err:  false,
+			want: false,
+		},
+		{
+			id: `?o = /u<peter>`,
+			in: []ConsumedElement{
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemBinding,
+					Text: "?o",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemEQ,
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemNode,
+					Text: "/u<peter>",
+				}),
+			},
+			r: table.Row{
+				"?o": &table.Cell{L: buildLiteralOrDie(`"73"^^type:int64`)},
+			},
+			err:  false,
+			want: false,
+		},
+		{
+			id: `?s ID ?id > "37"^^type:int64`,
+			in: []ConsumedElement{
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemBinding,
+					Text: "?id",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemGT,
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemLiteral,
+					Text: `"37"^^type:int64`,
+				}),
+			},
+			r: table.Row{
+				"?id": &table.Cell{S: table.CellString("peter")},
+			},
+			err:  true,
+			want: false,
+		},
+		{
+			id: `?s ID ?id = /u<peter>`,
+			in: []ConsumedElement{
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemBinding,
+					Text: "?id",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemEQ,
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemNode,
+					Text: "/u<peter>",
+				}),
+			},
+			r: table.Row{
+				"?id": &table.Cell{S: table.CellString("peter")},
+			},
+			err:  true,
+			want: false,
+		},
+		{
+			id: `?foo = "10"^^type:text`,
+			in: []ConsumedElement{
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemBinding,
+					Text: "?foo",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemEQ,
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemLiteral,
+					Text: `"10"^^type:text`,
+				}),
+			},
+			r: table.Row{
+				"?foo": &table.Cell{L: buildLiteralOrDie(`"10"^^type:int64`)},
+			},
+			err:  false,
+			want: false,
+		},
+		{
+			id: `?foo > "10"^^type:int64`,
+			in: []ConsumedElement{
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemBinding,
+					Text: "?foo",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemGT,
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemLiteral,
+					Text: `"10"^^type:int64`,
+				}),
+			},
+			r: table.Row{
+				"?foo": &table.Cell{L: buildLiteralOrDie(`"100.0"^^type:float64`)},
 			},
 			err:  false,
 			want: true,
 		},
 	}
 	for _, entry := range testTable {
 		eval, err := NewEvaluator(entry.in)
-		if !entry.err && err != nil {
```

Comment:
Please identify the function you are testing in the error messages, as well as the input and got/want.

For example, if you are testing the second return of `Evaluate` (the error) say:
`t.Errorf("e.Evaluate(%q) = _, %v; want nil", case.id, err)` // for a case where you didn't expect error but got one
And then when you proceed to comparing the first return:
`t.Errorf("e.Evaluate(%q) = %v, _; want %v, _", case.id, got, want)`
Note that these messages inform the maintainer what was the function under test, what was the input, the result you are looking at and what you expected. 

It would also be helpful, in this case, to print diffs so it is easier for the reader to compare got/want. For example:
`t.Errorf("e.Evaluate(%q) returned diff (-want +got):\n%s", case.id, diff)`
You can use a library such as `cmp.Diff` to do so:
https://godoc.org/github.com/google/go-cmp/cmp#Diff

Reply:
Done. Thank you very much for these insights Tati.